hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java

```diff
@@ -1168,4 +1169,43 @@ private V computeLocally(K key,
             }
         }
     }
+
+    public V merge(@Nonnull K key, @Nonnull V value,
+                   @Nonnull BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
+        checkNotNull(key, NULL_KEY_IS_NOT_ALLOWED);
+        checkNotNull(value, NULL_VALUE_IS_NOT_ALLOWED);
+        checkNotNull(key, NULL_BIFUNCTION_IS_NOT_ALLOWED);
+
+        if (SerializationUtil.isClassStaticAndSerializable(remappingFunction)
+                && isClusterVersionGreaterOrEqual(Versions.V4_1)) {
+            MergeEntryProcessor<K, V> ep = new MergeEntryProcessor<>(remappingFunction, value);
+            return executeOnKey(key, ep);
+        } else {
+            return mergeLocally(key, value, remappingFunction);
+        }
+    }
+
+    private V mergeLocally(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
+        while (true) {
+            Data oldValueAsData = toData(getInternal(key));
+            if (oldValueAsData != null) {
+                V oldValueClone = toObject(oldValueAsData);
+                V newValue = remappingFunction.apply(oldValueClone, value);
+                if (newValue != null) {
+                    if (replaceInternal(key, oldValueAsData, toData(newValue))) {
+                        return newValue;
+                    }
+                } else if (removeInternal(key, oldValueAsData)) {
+                    return null;
+                }
+            } else {
+                Data result =  putIfAbsentInternal(key, toData(value), UNSET, TimeUnit.MILLISECONDS, UNSET,
```

Comment:
I think we can avoid deserializing the key twice.
The deserialization happens inside function calls `getInternal` , `replaceInternal `, `putIfAbsentInternal`  and `removeInternal`  .
Instead, we can convert the key once, and pass the same data to all functions. These functions can work with data as the key. 

The same goes for ClientMapProxy. 
The only difference is that you need to use `toDataWithStrategy(key)` on MapProxyImpl for keys.